### ISSUE TITLE
Sketch of the PR

### ISSUE THREAD

**Issue by zhuzilin** [reyoung/torchrec#72]
Hi TorchRec Team!

After a few months of development, our team finished a PoC that adds dynamic embedding support on TorchRec and tested it on our production model. The main idea is to use GPU as the cache for the whole embedding, and store the other part of the embedding on PS. As feed ids are concentrated in most realistic scenarios, such design would make good use of GPU while giving no loss on the precision (AUC) of the model.

## API

As we proposed in the RFC, we add a `wrap` wrapper for the model and dataloader, which would create an asynchronous pipeline for the transformation of the feed ids, fetching of the weights and optimizer states, and eviction internally. An example snippet is:
```python
import torchrec_dynamic_embedding as tde

class Model(nn.Module):
    def __init__(self, config1, config2):
        super().__init__()
        self.emb1 = EmbeddingCollection(tables=config1, device=torch.device("meta"))
        self.emb2 = EmbeddingCollection(tables=config2, device=torch.device("meta"))
        ...

    def forward(self, kjt1, kjt2):
        ...

m = Model(config1, config2)
m = DistributedModelParallel(m)
dataloader = tde.wrap(
    "redis://127.0.0.1:6379/?prefix=model",
    dataloader,
    m,
    # configs of the embedding collections in the model
    { "emb1": config1, "emb2": config2 })

for label, kjt1, kjt2 in dataloader:
    output = m(kjt1, kjt2)
    ...
```

The last dict parameter can be removed if we can access the config from the shared modules.

Apart from `wrap`, we also add a `save` API to flush the GPU cache back to PS.

## ID Transformer
We provide 2 types of `IDTransformer` to transform the real feed ids, whose range would exceed the capacity of GPU (we call them `global_ids`), into the range of GPU caches (we call them `cache_ids`): `NaiveIDTransformer`, which use the `ska::flat_hash_map` to store the mapping and `CachelineTransformer`, which includes a highly customized open address hash table that fully utilizes of the CPU cache line. The performance of the 2 transformers is listed below:

- CPU Info:
  - Run on (16 X 4800 MHz CPU s)
  - CPU Caches:
    - L1 Data 32 KiB (x8)
    - L1 Instruction 32 KiB (x8)
    - L2 Unified 256 KiB (x8)
    - L3 Unified 16384 KiB (x1)
- Performance:
  | type of transformer | num cache embedding | range of global ids | number ids transformed / ms |
  |-|-|-|-|
  | naive |2e8|1e6 ~ 2e6|30.7k|
  | naive |2e8|1e10 ~ 2e10|12.2k|
  | cacheline |2e8|1e6 ~ 2e6|90.6k|
  | cacheline |2e8|1e10 ~ 2e10|23.0k|

The benchmarks could be reproduced with `naive_id_transformer_benchmark.cpp` and `cacheline_id_transformer_benchmark.cpp`.

## Evict Strategy
We provide a `MixedLXUStrategy` for evicting the old ids when the id transformer is full. `MixedLXUStrategy` uses a randomized algorithm similar to [the approximate LFU in redis](https://redis.io/docs/manual/eviction/) to combine both the timestamp and frequency of a global id within 32 bits. The performance of `MixedLXUStrategy` is listed below (CPU is the same as above):
- Update:
  | num of existed ids | num elements to transformer per iter | iteration iter (ms) |
  |-|-|-|
  |3e7|1M|14.7|
  |3e8|1M|14.7|

  (The transform speed is irrelevant to the number of existing ids).
- Eviction:
  | total of ids | num ids to evict | time (s) |
  |-|-|-|
  | 2e7 | 1e7 | 3.5 |
The benchmarks could be reproduced with `mixed_lfu_lru_strategy_benchmark.cpp` and `mixed_lfu_lru_strategy_evict_benchmark.cpp`.

## PS

The current implementation includes a Redis-based PS and a simple C API to register new PS to the library. One could use the `memory_io` in the test as an example to register a new PS. The user **does not** need to add special functionality support (e.g. updating parameter or averaging gradient) when registering their KV system as a PS extension.

## Install and Test

To install the current implementation:
```bash
cd contrib/dynamic_embedding
python setup.py develop
```
And the dynamic embedding will be installed as a separate package named `torchrec_dynamic_embedding`.

We incorporate `gtest` for the C++ code and use `unittest` for the python APIs. The tests make sure that the implementation does not have any precision loss. Please turn on the `TDE_WITH_TESTING` in `setup.py` to run tests. Note that for the python test, one needs to set the environment variable `TDE_MEMORY_IO_PATH` to the path of the compiled `memory_io.so`.

## Experiment on production models

We have tested the current implementation on our production model and observed no AUC loss on both Redis-based PS and simple memory PS. For a model of num embeddings 3e8 (number of ids to hold in embeddings), dynamic embedding enables us to run the model on pure GPU (no UVM involved) by setting the embedding size to 1e7, which improves performance from 7.6 iter/s to 9.0 iter/s.

Even when we still train with num embedding 3e8, the asynchronous pipeline and fast transformers make the performance loss on introducing dynamic embedding neglectable.

We also observed an AUC improvement of 0.3% when changing from `global_id % 3e8` to fit the `int64` global ids into the local machine to using dynamic embedding, a result of preventing any possible id mapping collision (different global id mapped into the same embedding).
